### DISCUSSION TITLE
Airflow 2.8.2 breaking change (#37396)

### DISCUSSION THREAD

**Discussion by NBardelot** [apache/airflow#38150] [Q&A]
# Context

We went on to patch from 2.8.1 to 2.8.2 in order to apply security fixes for [Airflow CVEs](https://www.cvedetails.com/vulnerability-list/vendor_id-45/product_id-52213/Apache-Airflow.html).

We've overlooked the fact that the [Airflow 2.8.2 release note](https://airflow.apache.org/docs/apache-airflow/stable/release_notes.html#airflow-2-8-2-2024-02-26) includes the following item:

> Remove markdown from core dependencies ([#37396](https://github.com/apache/airflow/pull/37396))

This change is a breaking change from a semver/ITIL standpoint (it does not patch an issue, it removes a feature). The issue is minor, but FYI it broke our requirements.

# Improvement

Dependencies removal should probably never be done without a deprecation warning please.

**Comment by Taragolis** [comment]
Could you elaborate a bit more which feature it is remove? This package do not use by Airflow itself.

**Comment by potiuk** [comment]
> This change is a breaking change from a semver/ITIL standpoint (it does not patch an issue, it removes a feature). The issue is minor, but FYI it broke our requirements.

No. It's not - you have quite wrong understanding what SemVer is - I suggets you read and digest https://semver.org/.

You can still install markdown if you want. No problem with that. Change in dependencies is clearly explained in semver as non-breaking - see https://semver.org/#faq 

Also having markdown installed is not a feature of Airflow that is explained in https://airflow.apache.org/docs/apache-airflow/stable/public-airflow-interface.html . 

Also you have to remember that SemVer is an intention, not a promise. And Semver Versioning is the intention that whatever we consider (and document) as Public interface is not going to break most user's workflow (based on public interface).

**Reply by potiuk** [reply]
Just for reference from SemVer FAQ:

```
What should I do if I update my own dependencies without changing the public API?

That would be considered compatible since it does not affect the public API. Software that explicitly depends on the same 
dependencies as your package should have their own dependency specifications and the author will notice any conflicts. 
Determining whether the change is a patch level or minor level modification depends on whether you updated your 
dependencies in order to fix a bug or introduce new functionality. We would usually expect additional code for the latter 
instance, in which case it’s obviously a minor level increment.
```

**Reply by NBardelot** [reply]
You have the wrong definition of API here. Some users (end users) will be interested in compatibility of http/REST APIs or public class interfaces, from a development standpoint. Yet, Airflow is also a platform, and users like me (platformers) will be interested in stability / compatibility from a deployment standpoint. As stated by Semver, the best practice is about dependencies specifications and API as public facing features (and similar concepts have long been used in technologies like Java under the best practice of Specification Version vs Implementation Version for example). The semantic describes a product's features, which for a platform like Airflow includes the deployement behaviour.

In my opinion, removing a feature - here a dependency - should not be done in a minor maintenance branch. Plus, as I understand #37396, it could have been done in the main branch, and ready for the next minor version. It does not seem it was tied to a fix included in the maintenance. Thus, from a platformer pespective like mine, it is a breaking change.

**Reply by potiuk** [reply]
Nope. You got it wrong. Our SemVer is for application users not for those who deploy the appand that how adopted SemVer - if you look at the GitHub Readme, our versioning rules ar e defined there. Look it up. This has been discussed in the past at devlist, agreed and voted upon (you can look it up in devlist archives). 

So while you have different opinion on the matter, the whole airflow community has a different one and it is confirmed by not only discussing it but also passing vote and documenting it in our versioning ryle and how we interpret our SemVer compatibility.

So your opinion against the opinion of the whole Airflow community sealed with a formal vote according to the Apache Software Foundation voting rules  https://www.apache.org/foundation/voting.html , not against **just**  my opinion. Look it up please before  claims that I am wrong. I can claim you are wrong in the context of our ASF - rule run project and whole community backing it with vote. You have just your opinion. 

But of course if you still think your opinion is right, then there are ways to change it - start devlist discussion, convince those who will take part in the discussion, propose change in our rules and pass a formal vote. We all follow the 'disagree but engage' rule here - so if you manage to convince the community and pas a vote, I will engage even if I disagree with it.

And while the current rules are agreed and voted and  documented, I suggest you also 'disagree but engage' - but contesting it without actual action to change the rules we currently operate under and attempting to convince others at devlist is just futile effort and only introduces noise

**Reply by NBardelot** [reply]
> I suggest you also 'disagree but engage'.

This is exactly the aim of my posting here as an Airflow github discussion / Q&A. I underline a context, an opinion, and propose an improvement, with detailed reasons. And by the way I have no problem with your disagreement. You're core to the project and I value your feedback.

I have witnessed first hand how difficult it was when the Docker image and Chart for example were not part of the project. Apache taking ownership of Airflow as a platform, and of the tooling required to make it a platform is a huge improvement, and thanks for that. And that's why I stand by my main point: changes around build dependencies and deployment have impact on platform integration and deployment. Now that you have taken ownership of those, people like me will be showing interest in stability.

**Reply by potiuk** [reply]
>  And by the way I have no problem with your disagreement. You're core to the project and I value your feedback.

This is disagreement of the whole community, not mine :). I am just explaining to you what project has agreed as a whole. Again by voting and documenting rules we follow.

>  underline a context, an opinion, and propose an improvement, with detailed reasons. 

Again. The only way to change our approach is to discus it on devlist and propose change to the rules we agree to. You are not discussing it with me now - you are discussing with rules we set, agreed to and voted. 

> changes around build dependencies and deployment have impact on platform integration and deployment. 

Yes, you are 100% right.. Current decision and agreeement is that yes, we know it might cause deployment issues and our SemVer promises do not cover those. Very delberately and reasons for that are document in those discussions you can search in devlist and our README in Github - where the whole chapter about versioning and our approach is explained (and we voted on it).

> Now that you have taken ownership of those, people like me will be showing interest in stability.

Airflow is an ASF for 7 years already, so I am not sure about taking ownership - it's been governed this way for as long as I remember, the rules around versioning have been coined over time and been gradually clarified and documented. But again -if you want to change the rules. Devlist is the only way - and I strongly suggest, before starting your discussion searching for past ones at the devlist (SemVer, Dependencies are good keywords) - our devlist (link in "community" section of our website) has really good search interface. You will find all the arguments and discussion there.

Basically - there is no point to discuss it here - it is important enough subject so that it has to be discussed at devlist.

**Reply by NBardelot** [reply]
> Airflow is an ASF for 7 years already.

I was referring ("those") to the Helm chart and the Docker image. They came to the Apache repo in the late 1.10 versions. 

I'll introduce the discussion to the devlist then, and ask for the community feedback. Thanks.
